qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -177,19 +175,27 @@ def _error(circuit, target, qubit):
     of circuit as a weak proxy for error.
     """
     if target is None:
-        return len(circuit)
+        if isinstance(circuit, list):
+            return len(circuit)
+        else:
+            return len(circuit._multi_graph) - 2
     else:
         if isinstance(circuit, list):
             gate_fidelities = [
                 1 - getattr(target[node.name].get((qubit,)), "error", 0.0) for node in circuit
             ]
         else:
             gate_fidelities = [
-                1 - getattr(target[inst.operation.name].get((qubit,)), "error", 0.0)
-                for inst in circuit
+                1 - getattr(target[inst.op.name].get((qubit,)), "error", 0.0)
+                for inst in circuit.op_nodes()
             ]
         gate_error = 1 - np.product(gate_fidelities)
         if gate_error == 0.0:
-            return -100 + len(circuit)  # prefer shorter circuits among those with zero error
+            if isinstance(circuit, list):
+                return -100 + len(circuit)
+            else:
+                return -100 + len(
+                    circuit._multi_graph
+                )  # prefer shorter circuits among those with zero error
```

Comment:
I still have a philosophical problem with this score discrimination, but in practice I think it's highly unlikely to ever come up, and it's unrelated to this PR anyway.

Reply:
I agree with you, it'll be better to fix it for real. I'll push a separate PR to fix it now to have it do a proper sort based on a tuple of `(error_score, sequence_len)`

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/9239